airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -162,6 +166,9 @@ def _evaluate_trigger_rule(
                     changed = ti.set_state(State.UPSTREAM_FAILED, session)
                 elif skipped:
                     changed = ti.set_state(State.SKIPPED, session)
+                elif removed and successes and ti.map_index > -1:
+                    if ti.map_index >= successes:
```

Comment:
why do we compare map index with number of upstream successes?  that seems odd?

Reply:
Hmm, yes now you mention it this feels like it's going to break in some other cases.
Like what if there is 1 mapped upstream which is in the failed state, one in the removed state, this would erroneously remove it I think?

Reply:
@ephraimbuddy Could you take another look at this PR/case please?

Reply:
> Hmm, yes now you mention it this feels like it's going to break in some other cases. Like what if there is 1 mapped upstream which is in the failed state, one in the removed state, this would erroneously remove it I think?


In this case, `successes` will be `0`, also failed=`1`, so the condition will not be reached and the taskinstance will be marked as `upstream_failed`. Same thing when we have skipped task instances. The condition to mark the task instance as removed will not be reached.

The condition for the task to be marked successful is if we have some `removed` task instances and `successful` task instances, `no failed`, `no skipped` and the task is mapped. So if we get here, if the `map_index` of the task instance is >= all successful task instances, it means the task instance upstream is removed because indexes go from -1 upwards, it's not possible to remove map_index 1 and still have map_index 3? 

If we have 5 mapped tasks(0,1,2,3,4), and we remove 2, we will have 3 mapped tasks(0,1,2). If these 3 are successful,(successes=3),  then the removed are those greater than or equal to the map index 3(3,4).

Reply:
What if a task has multiple upstreams?

`[a, b] >> mapped_task(list)` for instance?

Reply:
This would apply https://github.com/apache/airflow/blob/0c7b4cbf62925cf359648eff146f9f4b0c6e7775/airflow/ti_deps/deps/trigger_rule_dep.py#L165-L166 it won't get to line 169.

Line 169 is only satisfied if we have `removed`, `successes,` no `failed`, no `skipped` and mapped task

Reply:
Ahh good. This is important enough functionality (it's the very core of Airflow) that we should add atest cases covering things like this

Reply:
It does seem like it's covered here: https://github.com/apache/airflow/blob/7d6d182c53876898cca9c7ff0244de4a5cf7adb8/tests/models/test_taskinstance.py#L1065-L1216, I'm taking a look too

Reply:
Another option is removing this part altogether. It's not part of the deadlock issue but I feel that it's good to have stuff. 
My reason is this:
If at first run upstream was 3 and downstream was 3 too. Upstream created the downstream. We have 3 -> 3 successes.
Then we reduce upstream to 2, meaning one task is removed and we clear and rerun the dag, without this part of the change, we will end up running all 3 of the downstreams: upstream (2 successful, 1 removed). Downstream(3 successful)

Reply:
It's not this one specific line, they are all like this and that's the worry.

The test you highlighted doesn't use mapped tasks so I don't think it covers the case I highlighted.